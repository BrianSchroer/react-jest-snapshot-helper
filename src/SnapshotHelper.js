import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component Jest snapshot helper
 */
export default class SnapshotHelper {
  /**
   * Create new SnapshotHelper instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propsAdjustments) =>
      Object.assign({}, props, propsAdjustments);
  }

  /**
   * Fluent syntax helper to specify a callback function to be called by the test() function
   * to assign propsAdjustments to the props passed that were passed via the constructor.
   * (Defaults to assigning top-level props only.)
   */
  withPropsAdjuster = propsAdjuster => {
    this.adjustProps = propsAdjuster;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propsAdjustments,
   * then create JSON string snapshot for the component.
   */
  getSnapshot = propsAdjustments => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propsAdjustments);

    return renderer.create(<Component {...props} />).toJSON();
  };

  /**
   * Adjust props that were passed to the constructor with propsAdjustments, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = propsAdjustments => {
    expect(this.getSnapshot(propsAdjustments)).toMatchSnapshot();
  };

  /**
   * Get JSON string snapshot for the component.
   * @param {*} component - React component
   */
  static getSnapshot(component) {
    return new SnapshotHelper(component).getSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot.
   * @param {*} component - React component1
   */
  static test(component) {
    new SnapshotHelper(component).test();
  }
}
