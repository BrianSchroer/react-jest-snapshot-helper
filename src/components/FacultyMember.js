import React from 'react';

/* eslint-disable react/prop-types */

export default class FacultyMember extends React.Component {
  render() {
    const { name, role, speciality } = this.props;
    return (
      <div className="faculty-member">
        <p>name: {name}</p>
        <p>role: {role}</p>
        <div>
          <p>{speciality.subject}</p>
          <p>
            {speciality.accredited ? (
              <span>Accredited</span>
            ) : (
              <span>Not Accredited</span>
            )}
          </p>
        </div>
      </div>
    );
  }
}
