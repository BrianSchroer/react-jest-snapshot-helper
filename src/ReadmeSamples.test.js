import React from 'react';
import SnapshotHelper from '../src';
import FacultyMember from './components/FacultyMember';

const ROLE_TEACHER = 'x';
const ROLE_PRINCIPAL = 'x';
const ROLE_COUNSELOR = 'x';
const ROLE_TEACHING_ASSISTANT = 'x';

const defaultProps = {
  name: 'Test Name',
  role: ROLE_TEACHER,
  speciality: { subject: 'Math', accredited: true }
};

// constructor "remembers" component type and props
const snapshotHelper = new SnapshotHelper(
  (
    <FacultyMember
      name="Test Name"
      role={ROLE_TEACHER}
      speciality={{ subject: 'Math', accredited: true }}
    />
  )
);

const teacherSnapshot = snapshotHelper.getSnapshot();
const principalSnapshot = snapshotHelper.getSnapshot({ role: ROLE_PRINCIPAL });

describe('FacultyMember', () => {
  // test() without arguments tests snapshot for constructor-supplied props:
  it('should render ROLE_TEACHER properly', () => snapshotHelper.test());

  // test() with propAdjustments argument tests snapshot of new component
  // instance with adjustments assigned to original constructor-supplied props:
  it('should render ROLE_PRINCIPAL properly', () =>
    snapshotHelper.test({ name: 'Big Cheese', role: ROLE_PRINCIPAL }));
  it('should render ROLE_COUNSELOR properly', () =>
    snapshotHelper.test({ name: 'Deanna Troi', role: ROLE_COUNSELOR }));
  it('should render ROLE_TEACHING_ASSISTANT properly', () =>
    snapshotHelper.test({
      name: 'Snappy Helper',
      role: ROLE_TEACHING_ASSISTANT
    }));
});

const helper = new SnapshotHelper(
  <FacultyMember {...defaultProps} />
).withPropsAdjuster((props, adjustments) =>
  Object.assign({}, props, {
    speciality: Object.assign({}, props.speciality, adjustments)
  })
);

// ^ Override adjustor to override "child" specialty properties, not parent props:

describe('FacultyMember', () => {
  it('should render accredited English teacher properly', () => {
    helper.test({ subject: 'English', accredited: true });
  });
  it('should render unaccredited science teacher properly', () => {
    helper.test({ subject: 'Science', accredited: false });
  });
});

describe('FacultyMember', () => {
  const snapshot = SnapshotHelper.getSnapshot(
    <FacultyMember {...defaultProps} />
  );

  it('should render properly', () => {
    SnapshotHelper.test(<FacultyMember {...defaultProps} />);
  });
});
