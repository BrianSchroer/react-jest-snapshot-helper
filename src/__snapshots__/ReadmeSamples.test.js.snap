// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacultyMember should render ROLE_COUNSELOR properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Deanna Troi
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Math
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render ROLE_PRINCIPAL properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Big Cheese
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Math
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render ROLE_TEACHER properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Test Name
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Math
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render ROLE_TEACHING_ASSISTANT properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Snappy Helper
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Math
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render accredited English teacher properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Test Name
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      English
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Test Name
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Math
    </p>
    <p>
      <span>
        Accredited
      </span>
    </p>
  </div>
</div>
`;

exports[`FacultyMember should render unaccredited science teacher properly 1`] = `
<div
  className="faculty-member"
>
  <p>
    name: 
    Test Name
  </p>
  <p>
    role: 
    x
  </p>
  <div>
    <p>
      Science
    </p>
    <p>
      <span>
        Not Accredited
      </span>
    </p>
  </div>
</div>
`;
